File: src/main/java/edu/cnm/deepdive/sieve/array/Sieve.java
package edu.cnm.deepdive.sieve.array;

import java.util.Arrays;
import java.util.LinkedList;
import java.util.List;
import java.util.stream.IntStream;

public class Sieve {

  private static int[] sieve(int limit) {
    boolean[] candidates = new boolean[limit + 1];
    Arrays.fill(candidates, 2, limit + 1, true);
    for (int value = 2; value <= Math.sqrt(limit); value++) {
      if (candidates[value]) {
        for (int multiple = value * value; multiple <= limit; multiple += value) {
          candidates[multiple] = false;
        }
      }
    }
    return IntStream
        .range(0, candidates.length)
        .filter((value) -> candidates[value])
        .toArray();
  }

  public static void main(String... args) {
    int upperBound = 1_000_000;
    long start = System.currentTimeMillis();
    int[] primes = sieve(upperBound);
    long end = System.currentTimeMillis();
    System.out.printf("Java Sieve with boolean[] and int[]:%n"
            + "%1$,d primes found between %2$,d and %3$,d in %4$d ms.%n",
        primes.length, primes[0], primes[primes.length - 1], end - start);
  }

}

File: src/main/kotlin/edu/cnm/deepdive/sieve/array/sieve.kt
package edu.cnm.deepdive.sieve.array

import kotlin.math.sqrt

fun sieve(limit: Int): List<Int> {
    val primes = mutableListOf<Int>()
    if (limit >= 2) {
        val candidates = BooleanArray(2) { false } + BooleanArray(limit - 1) { true }
        for (factor in 2..sqrt(limit.toDouble()).toInt()) {
            if (candidates[factor]) {
                for (multiple in (factor * factor)..limit step factor) {
                    candidates[multiple] = false
                }
            }
        }
        candidates.forEachIndexed { number, isPrime ->
            if (isPrime) {
                primes.add(number)
            }
        }
    }
    return primes
}

fun main(args: Array<String>) {
    val start = System.currentTimeMillis()
    val upperBound = 1_000_000
    val primes = sieve(upperBound)
    val end = System.currentTimeMillis()
    println(
        """
            Kotlin Sieve with BooleanArray and List<Int>: 
            ${primes.size} primes found between ${primes.first()} and ${primes.last()} in ${end - start} ms.
        """.trimIndent()
    )
}